src/main/java/org/opentripplanner/middleware/controllers/api/TrackedTripController.java

```diff
@@ -0,0 +1,141 @@
+                    .withProduces(JSON_ONLY)
+                    .withResponses(SwaggerUtils.createStandardResponses()),
+                (request, response) -> trackTrip(request, TripStage.UPDATE), JsonUtils::toJson)
+            .post(path("/endtracking")
+                    .withDescription("Terminates the tracking of a monitored trip by the user.")
+                    .withProduces(JSON_ONLY)
+                    .withResponses(SwaggerUtils.createStandardResponses()),
+                (request, response) -> trackTrip(request, TripStage.END), JsonUtils::toJson);
+    }
+
+    /**
+     * Provide the correct response to the caller based on the trip stage.
+     */
+    private static Object trackTrip(Request request, TripStage tripStage) {
+        var payload = getTrackingPayloadFromRequest(request);
+        if (payload != null) {
+            switch (tripStage) {
+                case START:
+                    if (isTripAssociatedWithUser(request, payload.tripId)) {
+                        return ManageTripTracking.startTrackingTrip(payload);
+                    }
+                    break;
+                case UPDATE:
+                    return ManageTripTracking.updateTracking(
+                        payload,
+                        Objects.requireNonNull(getActiveJourney(request, payload)
+                    ));
+                case END:
+                    ManageTripTracking.endTracking(Objects.requireNonNull(getActiveJourney(request, payload)));
+                    break;
+                default:
+                    logMessageAndHalt(request, HttpStatus.BAD_REQUEST_400, "Unknown trip stage: " + tripStage);
+                    return null;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Get the expected tracking payload for the request. The populated contents of the payload will be determined by
+     * the trip stage.
+     */
+    private static TrackingPayload getTrackingPayloadFromRequest(Request request) {
+        TrackingPayload payload;
+        try {
+            payload = getPOJOFromRequestBody(request, TrackingPayload.class);
+        } catch (JsonProcessingException e) {
+            logMessageAndHalt(request, HttpStatus.BAD_REQUEST_400, "Error parsing JSON tracking payload.", e);
+            return null;
+        }
+        return payload;
+    }
+
+    /**
+     * Confirm that the monitored trip (that the user is on) belongs to them.
+     */
+    private static boolean isTripAssociatedWithUser(Request request, String tripId) {
+        var user = Auth0Connection.getUserFromRequest(request);
+
+        var monitoredTrip = Persistence.monitoredTrips.getById(tripId);
+        if (
+            monitoredTrip == null ||
+            (user.adminUser != null && !monitoredTrip.userId.equals(user.adminUser.id)) ||
+            (user.otpUser != null && !monitoredTrip.userId.equals(user.otpUser.id)) ||
+            (user.apiUser != null && !monitoredTrip.userId.equals(user.apiUser.id))
```

Comment:
I think only `OtpUsers` can have monitored trips. An `ApiUser` or an `AdminUser` could create and links trips on behalf of other `OtpUser`s but I don't see a case where they alone would monitor trips, unless they are also `OtpUser`s.

Reply:
That does look the case so have addressed in https://github.com/ibi-group/otp-middleware/pull/203/commits/ec0152118d75a99b1cdf92e774c3e088927e6273. I can revert later if need be.